app/services/listing_index_service/search/zappy_adapter.rb

```diff
@@ -0,0 +1,84 @@
+module ListingIndexService::Search
+
+  class ZappyAdapter < SearchEngineAdapter
+
+    INCLUDE_MAP = {
+      listing_images: :listing_images,
+      author: :author,
+      num_of_reviews: {author: :received_testimonials},
+      location: :location
+    }
+
+    def initialize
+      @conn = Faraday.new(url: "http://127.0.0.1:8080") do |c|
+         c.request  :url_encoded             # form-encode POST params
+         c.response :logger                  # log requests to STDOUT
+         c.response :json, :content_type => /\bjson$/
+         c.adapter  Faraday.default_adapter  # make requests with Net::HTTP
+         c.use Faraday::Response::RaiseError if Rails.env.development?
+      end
+    end
+
+    def search(community_id:, search:, includes: nil)
+      included_models = includes.map { |m| INCLUDE_MAP[m] }
+      search_params = format_params(search)
+
+      if needs_db_query?(search) && needs_search?(search)
+        return Result::Error.new(ArgumentError.new("Both DB query and search engine would be needed to fulfill the search"))
+      end
+
+      if needs_search?(search)
+        # TODO: is out-of-bounds check necessary here?
+        begin
+          res = @conn.get do |req|
+            req.url("/api/v1/marketplace/#{community_id}/listings", search_params)
+            req.headers['Authorization'] = 'apikey key=asdfasdf'
+          end.body
+          Result::Success.new(parse_response(res, includes))
+        rescue StandardError => e
+          Result::Error.new(e)
+        end
+      else
+        fetch_from_db(community_id: community_id,
+                      search: search,
+                      included_models: included_models,
+                      includes: includes)
+      end
+    end
+
+    private
+
+    def format_params(original)
+      defaults = {
+        include_closed: false
+      }
+      params = {}
+      params[:'page[number]'] = original[:page] if original[:page]
+      params[:'page[size]'] = original[:per_page] if original[:per_page]
+      params[:keywords] = original[:keywords] if original[:keywords]
+      params[:include_closed] = original[:include_closed] if original[:include_closed]
```

Comment:
This pattern is used somewhere in our codebase, however, I find it a bit complex. It has a lot of mutation going on and a lot of if-branches.

How about using `compact?`

```ruby
{
  :include_closed => false,
  :'page[number]' => original[:page],
  :'page[size]' => original[:per_page],
  ...
}.compact
```

(of course, this doesn't work if `nil` actually means something)